src/session-pool.js

```diff
@@ -550,399 +288,178 @@ SessionPool.prototype.createTransaction_ = function(session, options) {
- * may try to use a different type.
- *
- * @private
+ * Make an API request, first assuring an active session is used.
  *
- * @param {string} type The desired session type.
- * @returns {Promise<Session>}
+ * @param {object} config
+ * @param {function} callback
  */
-SessionPool.prototype.getSession_ = function(type) {
-  var self = this;
-
-  var available = this.available();
-  var acquires = this.acquireQueue_.size;
-
-  if (available && !acquires) {
-    return this.getNextAvailableSession_(type);
-  }
-
-  if (this.options.fail) {
-    return Promise.reject(new Error('No resources available.'));
-  }
+SessionPool.prototype.request = function(config, callback) {
+  const self = this;
 
-  var promises = [this.waitForNextAvailable_(type)];
-  var shouldCreate =
-    !this.isFull() && available + this.pendingCreates_ < acquires + 1;
+  this.getReadSession().then(function(session) {
+    config.reqOpts.session = session.formattedName_;
 
-  if (shouldCreate) {
-    var createPromise = new Promise(function(resolve, reject) {
-      self.createSession_(type).then(function() {
-        self.emit('available');
-      }, reject);
+    self.request_(config, function() {
+      self.release(session);
+      callback.apply(null, arguments);
     });
-
-    promises.push(createPromise);
-  }
-
-  return this.race_(promises);
-};
-
-/**
- * Get the corresponding session group.
- *
- * @private
- *
- * @param {Session} session The session to get the group for.
- * @returns {Session[]}
- */
-SessionPool.prototype.getSessionGroup_ = function(session) {
-  return session.type === READWRITE ? this.writes_ : this.reads_;
-};
-
-/**
- * Indicates whether or not the pool needs to be filled.
- *
- * @private
- *
- * @return {boolean}
- */
-SessionPool.prototype.needsFill_ = function() {
-  return this.pendingCreates_ + this.size() < this.options.min;
-};
-
-/**
- * Creates a promise that will settle once a Session becomes available.
- *
- * @private
- *
- * @returns {Promise}
- */
-SessionPool.prototype.onAvailable_ = function() {
-  var self = this;
-
-  return new Promise(function(resolve) {
-    self.once('available', resolve);
-  });
+  }, callback);
 };
 
 /**
- * Makes a keep alive request to all the idle sessions.
- *
- * @private
- *
- * @fires SessionPool#error
- * @returns {Promise}
+ * This method will give you stats for the read and write sessions
+ * @returns {Object} stats
  */
-SessionPool.prototype.pingIdleSessions_ = function() {
-  var self = this;
-  var sessions = this.getIdleSessions_();
-
-  return Promise.all(
-    sessions.map(function(session) {
-      return self.pingSession_(session);
-    })
-  ).then(function() {
-    if (self.needsFill_()) {
-      self.fill();
-    }
-  });
+SessionPool.prototype.getStats = function() {
+  const readPool = this.readPool;
+  const writePool = this.writePool;
+  const stats = {
+    readPool: {
+      spareResourceCapacity: readPool.spareResourceCapacity,
```

Comment:
How big is this object, why wouldnt you just do an `object.assign` here?

Reply:
That object is tiny, just a few numbers.